File: clients/tv-native/modules/vlc-player/android/src/main/java/expo/modules/vlcplayer/VlcPlayerView.kt
package expo.modules.vlcplayer

import android.content.Context
import android.view.SurfaceView
import android.view.ViewGroup
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import org.videolan.libvlc.LibVLC
import org.videolan.libvlc.Media
import org.videolan.libvlc.MediaPlayer
import org.videolan.libvlc.util.VLCVideoLayout

// `--no-sub-autodetect-file`: KROMA owns subtitle selection and draws it in React,
// so VLC must never pick up a sidecar. The embedded tracks are disabled separately,
// on the player (`spuTrack = -1`), since this option only covers files on disk.
// `--network-caching`: a remote HLS master needs more than the 1s LAN default or
// the first seek stalls.
// The chrome interpolates between reports, so a coarser interval than the frame
// rate is plenty and costs less. Matches the expo-video plane's own cap.
private const val TIME_EMIT_MS = 250L

private val VLC_ARGS =
  arrayListOf(
    "--no-sub-autodetect-file",
    "--network-caching=1500",
    "--audio-time-stretch",
  )

class VlcPlayerView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  private val onPlayerTime by EventDispatcher()
  private val onPlayerLoad by EventDispatcher()
  private val onPlayerState by EventDispatcher()
  private val onPlayerError by EventDispatcher()

  private val videoLayout = VLCVideoLayout(context)
  private var libVlc: LibVLC? = null
  private var player: MediaPlayer? = null
  private var attached = false
  private var pendingSeekMs: Long? = null
  private var startMs: Long = 0
  private var loadedUri: String? = null
  private var pendingUri: String? = null
  private var seekTargetMs: Long = 0
  private var lastSeekNonce = -1
  private var wantPaused = false
  private var wantAudioIndex = 0
  private var audioFilter = "off"
  private var rate: Float = 1f
  private var lastTimeEmit = 0L

  init {
    videoLayout.layoutParams =
      ViewGroup.LayoutParams(
        ViewGroup.LayoutParams.MATCH_PARENT,
        ViewGroup.LayoutParams.MATCH_PARENT,
      )
    addView(videoLayout)
  }

  private fun ensurePlayer(): MediaPlayer {
    player?.let { return it }
    val vlc = LibVLC(context, VLC_ARGS)
    val mp = MediaPlayer(vlc)
    mp.setEventListener { event ->
      when (event.type) {
        MediaPlayer.Event.TimeChanged -> {
          // Gated here, not in JS: libVLC reports at the demuxer's PCR rate, several
          // times faster than the chrome's budget, and every crossing re-renders the
          // whole player tree. The event already carries the time, so the gate also
          // saves a JNI read per tick. Duration arrives on LengthChanged.
          val now = android.os.SystemClock.uptimeMillis()
          if (now - lastTimeEmit >= TIME_EMIT_MS) {
            lastTimeEmit = now
            onPlayerTime(mapOf("timeMs" to event.timeChanged))
          }
        }
        MediaPlayer.Event.LengthChanged -> onPlayerLoad(mapOf("lengthMs" to event.lengthChanged))
        MediaPlayer.Event.ESAdded -> {
          applyAudioTrack()
          if (mp.spuTrack != -1) mp.spuTrack = -1
        }
        MediaPlayer.Event.Playing -> {
          // The resume point, now that the demux can answer a seek. Applied here
          // rather than from JS so it costs no bridge round trip.
          pendingSeekMs?.let {
            pendingSeekMs = null
            mp.time = it
          }
          applyAudioTrack()
          // KROMA draws subtitles in React over the plane. Left to itself VLC
          // burns the embedded track into the picture, and the viewer gets two.
          mp.spuTrack = -1
          mp.volume = boostVolume()
          mp.rate = rate
          onPlayerState(mapOf("state" to "playing"))
        }
        MediaPlayer.Event.Paused -> onPlayerState(mapOf("state" to "paused"))
        MediaPlayer.Event.Buffering ->
          onPlayerState(mapOf("state" to "buffering", "percent" to event.buffering))
        MediaPlayer.Event.EndReached -> onPlayerState(mapOf("state" to "ended"))
        MediaPlayer.Event.EncounteredError ->
          onPlayerError(mapOf("message" to "VLC could not open or decode this stream"))
      }
    }
    libVlc = vlc
    player = mp
    if (!attached) {
      mp.attachViews(videoLayout, null, false, false)
      attached = true
    }
    return mp
  }

  fun setStartMs(ms: Long) {
    startMs = ms
  }

  // Kept on the view, because a new MediaPlayer starts at 1x: a rate chosen before
  // the stream opened (or before a filter reopened it) has to survive both.
  fun setRate(next: Float) {
    rate = next
    player?.rate = next
  }

  fun setSource(uri: String?) {
    pendingUri = uri
  }

  // Opening is deferred to the end of the prop batch, never done inside a setter:
  // Expo applies props in declaration order, so a filter or an offset that sorts
  // after the URL would otherwise open the stream with the previous value and then
  // reopen it - losing the resume point on the way. Idempotent, because props
  // re-apply on every render and reopening would restart the film under the viewer.
  fun commit() {
    val uri = pendingUri
    if (uri.isNullOrEmpty() || uri == loadedUri) return
    loadedUri = uri
    load(uri, startMs)
  }

  private fun load(uri: String, startMs: Long) {
    try {
      val vlcPlayer = ensurePlayer()
      val vlc = libVlc ?: return
      val media = Media(vlc, android.net.Uri.parse(uri))
      // Enabled but NOT forced: forcing removes the software fallback, and a chip
      // that cannot take the stream then stalls on black. The order is spelled out
      // rather than left to VLC's ranking - software is the fallback, never the
      // default, since a TV SoC cannot sustain it.
      media.setHWDecoderEnabled(true, false)
      media.addOption(":codec=mediacodec_ndk,mediacodec_jni,avcodec")
      for (option in filterOptions()) media.addOption(option)
      // NOT `:start-time`: that is a demux-level skip, and on a container VLC
      // cannot seek by time it reads from byte zero and throws frames away, which
      // over HTTP means pulling the whole file to reach the resume point. The
      // offset is applied as a real seek on the first Playing event instead, which
      // libVLC serves with a Range request.
      vlcPlayer.media = media
      pendingSeekMs = if (startMs > 0) startMs else null
      if (!wantPaused) vlcPlayer.play()
      // Released only AFTER play(): the player retains it, but releasing first is
      // the one link in this chain that was never proven on this libVLC build.
      media.release()
    } catch (t: Throwable) {
      onPlayerError(mapOf("message" to (t.message ?: "VLC could not open this stream")))
    }
  }


  // VLC's compressor takes dB where the server's ffmpeg chain takes a linear
  // threshold, so these are the same two curves converted, not new ones:
  // 0.063 -> -24 dB and 0.04 -> -28 dB, makeup 1.4x -> about +3 dB. `boost` is
  // not a compressor at all - it is gain past 100%, for a track that is simply
  // recorded too quiet.
  private fun filterOptions(): List<String> =
    when (audioFilter) {
      "standard" ->
        listOf(
          ":audio-filter=compressor",
          ":compressor-threshold=-24",
          ":compressor-ratio=4",
          ":compressor-attack=10",
          ":compressor-release=250",
          ":compressor-knee=6",
          ":compressor-makeup-gain=3",
        )
      "night" ->
        listOf(
          ":audio-filter=compressor",
          ":compressor-threshold=-28",
          ":compressor-ratio=8",
          ":compressor-attack=4",
          ":compressor-release=250",
          ":compressor-knee=5",
          // No make-up: night is the quietest mode by design, and every other copy of
          // this curve trims BELOW unity (ffmpeg `volume=0.9`). VLC cannot go under
          // unity here, so zero is as close as the filter reaches.
          ":compressor-makeup-gain=0",
        )
      else -> emptyList()
    }

  // 100 is unity; libVLC amplifies up to 200. Applied on every start too, since a
  // new MediaPlayer comes up at unity.
  private fun boostVolume(): Int = if (audioFilter == "boost") 175 else 100

  // A filter belongs to the media, so changing one reopens the stream where it
  // stands. That costs one byte-range request, not a re-download, and it is the
  // only way libVLC takes a new filter chain.
  fun setAudioFilter(mode: String) {
    if (mode == audioFilter) return
    val wasBoostOnly = audioFilter == "boost" || audioFilter == "off"
    audioFilter = mode
    val mp = player
    if (mp != null && (mode == "boost" || mode == "off") && wasBoostOnly) {
      mp.volume = boostVolume()
      return
    }
    val uri = loadedUri ?: return
    startMs = mp?.time ?: 0
    loadedUri = null
    pendingUri = uri
    commit()
  }

  fun setPaused(paused: Boolean) {
    wantPaused = paused
    val mp = player ?: return
    if (paused) {
      if (mp.isPlaying) mp.pause()
    } else if (!mp.isPlaying) {
      mp.play()
    }
  }

  fun setSeekTarget(ms: Long) {
    seekTargetMs = ms
  }

  // Props re-apply on every render, so the nonce is what separates a new request
  // from the same one arriving again.
  fun applySeek(nonce: Int) {
    if (nonce == lastSeekNonce) return
    lastSeekNonce = nonce
    if (nonce <= 0) return
    seekTo(seekTargetMs)
  }


  // VLC rejects a seek before the media is parsed, so a seek arriving early is
  // held and applied on the first Playing event.
  fun seekTo(ms: Long) {
    val mp = player ?: return
    if (mp.length <= 0) {
      pendingSeekMs = ms
      return
    }
    mp.time = ms
  }


  // KROMA counts audio tracks in file order; VLC keys them by its own id. The index
  // is remembered because VLC has no track list until the media is parsed, and a
  // language chosen while the stream opens would otherwise be dropped.
  fun setAudioTrack(index: Int) {
    if (index < 0) return
    wantAudioIndex = index
    applyAudioTrack()
  }

  private fun applyAudioTrack() {
    val index = wantAudioIndex
    val mp = player ?: return
    val playable = mp.audioTracks?.filter { it.id >= 0 } ?: return
    val track = playable.getOrNull(index) ?: return
    if (mp.audioTrack != track.id) mp.audioTrack = track.id
  }

  fun release() {
    player?.let {
      it.stop()
      if (attached) {
        it.detachViews()
        attached = false
      }
      it.release()
    }
    player = null
    libVlc?.release()
    libVlc = null
    // Cleared with the player: a remounted view is asked for the same URL, and a
    // stale value makes commit() treat it as already open and draw nothing.
    loadedUri = null
  }

  // React Native lays out ITS OWN views and stops there: a child added natively is
  // never measured, so it keeps a zero-sized box, its SurfaceView never produces a
  // surface, and libVLC waits for one before it will open the input. Audio comes up
  // (it needs no window), the picture never does, and nothing errors.
  override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
    val w = right - left
    val h = bottom - top
    videoLayout.measure(
      MeasureSpec.makeMeasureSpec(w, MeasureSpec.EXACTLY),
      MeasureSpec.makeMeasureSpec(h, MeasureSpec.EXACTLY),
    )
    videoLayout.layout(0, 0, w, h)
  }

  override fun onDetachedFromWindow() {
    release()
    super.onDetachedFromWindow()
  }
}

File: clients/tv-native/modules/vlc-player/ios/VlcPlayerView.swift
import ExpoModulesCore

#if os(tvOS)
  import TVVLCKit
#else
  import MobileVLCKit
#endif

// The Apple half of the libVLC plane; see the Kotlin next door for Android. The
// reason it exists is sharper here than there: AVFoundation has no Matroska
// demuxer at all, so today every MKV is a server remux. VLC carries its own
// demuxers and decoders, which lets those titles direct-play instead.
final class VlcPlayerView: ExpoView {
  private let onPlayerTime = EventDispatcher()
  private let onPlayerLoad = EventDispatcher()
  private let onPlayerState = EventDispatcher()
  private let onPlayerError = EventDispatcher()

  private let surface = UIView()
  private var player: VLCMediaPlayer?
  private var loadedUri: String?
  private var pendingUri: String?
  private var startMs: Int64 = 0
  private var pendingSeekMs: Int64?
  private var seekTargetMs: Int64 = 0
  private var lastSeekNonce = -1
  private var wantPaused = false
  private var wantAudioIndex = 0
  private var audioFilter = "off"
  private var rate: Float = 1
  private var lastLengthMs: Int64 = 0
  private var lastTimeEmit: TimeInterval = 0

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = true
    addSubview(surface)
  }

  // The plane fills the view; React lays out ITS own views only, so a subview
  // added here is positioned by hand or it stays at zero and never draws.
  override func layoutSubviews() {
    super.layoutSubviews()
    surface.frame = bounds
  }

  private func ensurePlayer() -> VLCMediaPlayer {
    if let existing = player { return existing }
    // `--no-sub-autodetect-file`: KROMA draws subtitles in React, so VLC must not
    // pick up a sidecar. Embedded tracks are disabled separately, on the player.
    let media = VLCMediaPlayer(options: [
      "--no-sub-autodetect-file",
      "--network-caching=1500",
      "--audio-time-stretch",
    ])
    media.drawable = surface
    media.delegate = self
    player = media
    return media
  }

  // VLC's compressor takes dB where the server's ffmpeg chain takes a linear
  // threshold: the same two curves converted, not new ones. `boost` is gain past
  // unity rather than compression, for a track recorded too quietly.
  private func filterOptions() -> [String] {
    switch audioFilter {
    case "standard":
      return [
        ":audio-filter=compressor", ":compressor-threshold=-24", ":compressor-ratio=4",
        ":compressor-attack=10", ":compressor-release=250", ":compressor-knee=6",
        ":compressor-makeup-gain=3",
      ]
    case "night":
      return [
        ":audio-filter=compressor", ":compressor-threshold=-28", ":compressor-ratio=8",
        ":compressor-attack=4", ":compressor-release=250", ":compressor-knee=5",
        // No make-up: night is the quietest mode by design; see the Kotlin twin.
        ":compressor-makeup-gain=0",
      ]
    default:
      return []
    }
  }

  private func boostVolume() -> Int32 { audioFilter == "boost" ? 175 : 100 }

  func setStartMs(_ ms: Int64) { startMs = ms }

  // Kept on the view, because a new player starts at 1x: a rate chosen before the
  // stream opened (or before a filter reopened it) has to survive both.
  func setRate(_ next: Float) {
    rate = next
    player?.rate = next
  }

  func setSource(_ uri: String?) {
    pendingUri = uri
  }

  // Opening is deferred to the end of the prop batch, never done inside a setter:
  // Expo applies props in declaration order, so a filter or an offset that sorts
  // after the URL would otherwise open the stream with the previous value and then
  // reopen it - losing the resume point on the way. Idempotent, because props
  // re-apply on every render and reopening would restart the film under the viewer.
  func commit() {
    guard let uri = pendingUri, !uri.isEmpty, uri != loadedUri, let url = URL(string: uri)
    else { return }
    loadedUri = uri
    // Reset per open: it gates the pending seek, and a reopen of the same media has
    // an identical length, so a stale value swallows the resume point forever.
    lastLengthMs = 0
    let vlc = ensurePlayer()
    let media = VLCMedia(url: url)
    // Spelled out rather than left to VLC's own ranking: VideoToolbox first, and
    // avcodec only as the last resort. Software decoding is what this engine
    // exists for, but it is the fallback, never the default.
    media.addOption(":codec=videotoolbox,avcodec")
    for option in filterOptions() { media.addOption(option) }
    vlc.media = media
    pendingSeekMs = startMs > 0 ? startMs : nil
    if !wantPaused { vlc.play() }
  }

  func setPaused(_ paused: Bool) {
    wantPaused = paused
    guard let vlc = player else { return }
    if paused {
      if vlc.isPlaying { vlc.pause() }
    } else if !vlc.isPlaying {
      vlc.play()
    }
  }

  func setSeekTarget(_ ms: Int64) { seekTargetMs = ms }

  // Props re-apply on every render, so the nonce is what separates a new request
  // from the same one arriving again.
  func applySeek(_ nonce: Int) {
    guard nonce != lastSeekNonce else { return }
    lastSeekNonce = nonce
    guard nonce > 0 else { return }
    seek(to: seekTargetMs)
  }

  private func seek(to ms: Int64) {
    guard let vlc = player else { return }
    guard lastLengthMs > 0 else {
      pendingSeekMs = ms
      return
    }
    vlc.time = VLCTime(number: NSNumber(value: ms))
  }

  // KROMA counts audio tracks in file order; VLC keys them by its own id, so an
  // index cannot be handed over as-is. The wanted index is remembered because
  // VLC has no track list until the media is parsed.
  func setAudioTrack(_ index: Int) {
    guard index >= 0 else { return }
    wantAudioIndex = index
    applyAudioTrack()
  }

  private func applyAudioTrack() {
    guard let vlc = player, let ids = vlc.audioTrackIndexes as? [NSNumber] else { return }
    let playable = ids.filter { $0.intValue >= 0 }
    guard wantAudioIndex < playable.count else { return }
    let wanted = playable[wantAudioIndex].int32Value
    if vlc.currentAudioTrackIndex != wanted { vlc.currentAudioTrackIndex = wanted }
  }

  // A filter belongs to the media, so changing one reopens the stream where it
  // stands: one byte-range request, not a re-download. Boost is only a volume
  // change and is applied in place.
  func setAudioFilter(_ mode: String) {
    guard mode != audioFilter else { return }
    let wasGainOnly = audioFilter == "boost" || audioFilter == "off"
    audioFilter = mode
    if let vlc = player, wasGainOnly, mode == "boost" || mode == "off" {
      vlc.audio?.volume = boostVolume()
      return
    }
    guard let uri = loadedUri else { return }
    startMs = Int64(player?.time.intValue ?? 0)
    loadedUri = nil
    pendingUri = uri
    commit()
  }

  func release() {
    player?.stop()
    player?.delegate = nil
    player = nil
    // Cleared with the player: a remounted view is asked for the same URL, and a
    // stale value makes commit() treat it as already open and draw nothing.
    loadedUri = nil
    lastLengthMs = 0
  }

  override func removeFromSuperview() {
    release()
    super.removeFromSuperview()
  }
}

extension VlcPlayerView: VLCMediaPlayerDelegate {
  func mediaPlayerTimeChanged(_: Notification) {
    guard let vlc = player else { return }
    let lengthMs = Int64(vlc.media?.length.intValue ?? 0)
    if lengthMs > 0, lengthMs != lastLengthMs {
      lastLengthMs = lengthMs
      onPlayerLoad(["lengthMs": lengthMs])
      if let pending = pendingSeekMs {
        pendingSeekMs = nil
        vlc.time = VLCTime(number: NSNumber(value: pending))
      }
    }
    // Gated: VLC reports at the demuxer's rate, several times faster than the
    // chrome's budget, and every crossing re-renders the whole player tree.
    let now = ProcessInfo.processInfo.systemUptime
    guard now - lastTimeEmit >= 0.25 else { return }
    lastTimeEmit = now
    onPlayerTime(["timeMs": Int64(vlc.time.intValue)])
  }

  func mediaPlayerStateChanged(_: Notification) {
    guard let vlc = player else { return }
    switch vlc.state {
    case .playing:
      applyAudioTrack()
      // KROMA draws subtitles in React over the plane; left alone VLC burns the
      // embedded track into the picture and the viewer gets two.
      vlc.currentVideoSubTitleIndex = -1
      vlc.audio?.volume = boostVolume()
      vlc.rate = rate
      onPlayerState(["state": "playing"])
    case .paused:
      onPlayerState(["state": "paused"])
    case .buffering:
      onPlayerState(["state": "buffering", "percent": 100])
    case .ended, .stopped:
      onPlayerState(["state": "ended"])
    case .error:
      onPlayerError(["message": "VLC could not open or decode this stream"])
    default:
      break
    }
  }
}
